agent/multi-recorder-agent.c

```diff
@@ -0,0 +1,224 @@
+      hwangsae_recorder_agent_get_recorder_id (recorder_agent);
+  guint port = hwangsae_recorder_agent_get_relay_stream_port (recorder_agent);
+  g_autofree gchar *recording_dir =
+      hwangsae_recorder_agent_get_recording_dir (recorder_agent);
+  g_autofree gchar *streamid_tmp = NULL;
+  g_autofree gchar *streamid = NULL;
+  g_autofree gchar *url = NULL;
+  g_autofree gchar *recording_edge_dir = NULL;
+  g_autofree gchar *filename_prefix = NULL;
+  RecordingData *recording_data;
+  gchar *edge_id_key;
+
+  LOCK_MULTI_RECORDER_AGENT;
+
+  recording_data = g_hash_table_lookup (self->edge_map, edge_id);
+  if (recording_data) {
+    g_warning ("recording already started");
+    return recording_data->recording_id;
+  }
+
+  recording_data = g_new0 (RecordingData, 1);
+  recording_data->recording_id = g_get_real_time ();
+  recording_data->recorder = hwangsae_recorder_new ();
+
+  hwangsae_recorder_agent_send_rest_api (recorder_agent,
+      RELAY_METHOD_START_STREAMING, edge_id);
+
+  streamid_tmp = g_strdup_printf ("#!::r=%s,u=%s", edge_id, recorder_id);
+  streamid = g_uri_escape_string (streamid_tmp, NULL, FALSE);
+  url = g_strdup_printf ("srt://%s:%d?streamid=%s", host, port, streamid);
+
+  g_debug ("starting to recording stream from %s", url);
+
+  if (g_str_equal (recording_dir, "")) {
+    g_free (recording_dir);
+    recording_dir = g_build_filename (g_get_user_data_dir (),
+        "hwangsaeul", "hwangsae", "recordings", NULL);
+  }
+
+  recording_edge_dir = g_build_filename (recording_dir, edge_id, NULL);
+
+  filename_prefix = g_strdup_printf ("hwangsae-recording-%ld",
+      recording_data->recording_id);
+
+  g_debug ("setting recording_dir: %s, filename_prefix: %s\n",
+      recording_edge_dir, filename_prefix);
+
+  hwangsae_recorder_set_recording_dir (recording_data->recorder,
+      recording_edge_dir);
+
+  hwangsae_recorder_set_filename_prefix (recording_data->recorder,
+      filename_prefix);
+
+  hwangsae_recorder_set_container (recording_data->recorder,
+      HWANGSAE_CONTAINER_TS);
+
+  g_signal_connect (recording_data->recorder, "stream-disconnected",
+      (GCallback) stream_disconnected_cb, self);
+
+  edge_id_key = g_strdup (edge_id);
+  g_hash_table_insert (self->edge_map, edge_id_key, recording_data);
+
+  hwangsae_recorder_start_recording (recording_data->recorder, url);
+
+  return recording_data->recording_id;
+}
+
+static void
+hwangsae_multi_recorder_agent_stop_recording (HwangsaeRecorderAgent *
+    recorder_agent, gchar * edge_id)
+{
+  g_autoptr (GError) error = NULL;
+  RecordingData *recording_data;
+  HwangsaeMultiRecorderAgent *self =
+      HWANGSAE_MULTI_RECORDER_AGENT (recorder_agent);
+
+  LOCK_MULTI_RECORDER_AGENT;
+
+  recording_data = g_hash_table_lookup (self->edge_map, edge_id);
+
+  if (!recording_data) {
+    g_warning ("recording already stopped");
+    return;
+  }
+
+  hwangsae_recorder_stop_recording (recording_data->recorder);
+
+  g_hash_table_remove (self->edge_map, edge_id);
```

Comment:
Are you silently transferring ownership of the recorder reference from the hash table so that it gets unref'ed in `stream_disconnected_cb`?

I think it would be more clear to explicitly `g_object_ref()` the recorder here (and maybe add comment where the reference will get freed) and let `_recording_data_free()` clear the original reference when the item gets removed from the map.

Otherwise you're leaking recorders in `hwangsae_multi_recorder_agent_dispose()`.

Reply:
Yes, the `recorder` object gets clear on `stream_disconnected_cb`, I pointed the issue in 
https://github.com/hwangsaeul/hwangsae/pull/55#discussion_r356366988

Basically we need to wait for the `recorder` to finish its internal `cleanup` after `hwangsae_recorder_stop_recording` and call `hwangsae_recorder_stop_recording_internal`, which is the one that triggers the stream_disconnected_cb`.

Please clarify if you understand there is a better way to implement this.

Reply:
I understand we have to wait for "stream-disconnected" before we release the final reference. I only wanted to make the rather less obvious passage of the reference more explicit in the code, e.g.

```
/* Will get released in stream_disconnected_cb. */
g_object_ref (recording_data->recorder);

hwangsae_recorder_stop_recording (recording_data->recorder);

g_hash_table_remove (self->edge_map, edge_id); (this calls _recording_data_free which ALSO removes its own recorder reference)
```

> Please clarify if you understand there is a better way to implement this.

Hmm, one way to make the API better might be implementing it as an asynchronous operation using `GTask` ([here](https://developer.gnome.org/gio/stable/GTask.html)). Then you'd have

```
void
hwangsae_recorder_stop_recording (HwangsaeRecorder    *self,
                                  GCancellable        *cancellable,
                                  GAsyncReadyCallback  callback,
                                  gpointer             user_data)
```
Finishing the operation would invoke `callback`, passing `user_data`. In this case the callback would just unref the recorder as `stream_disconnected_cb` does , but without reliance on an inobvious relation between `hwangsae_recorder_stop_recording()` and `"stream-disconnected"` signal.

Reply:
Thanks for sharing this information and clarify. I have added the reference and comment as you suggested to the code.

Reply:
... but now you have one reference too many because you're not releasing recorder in `_recording_data_free()`, don't you?

Reply:
Yes, you are correct.